File: C4_InnerTube/InnerTubeCS/SharedUtilities/FileHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;

namespace SharedUtilities
{
    public enum FileType
    {
        Image,
        Flv,
        Mp4,
        Wmv
    }

    public static class FileHelper
    {
        public static string DefaultAppName = "InnerTube";
        public static string DefaultSettingsName = "settings.xml";
        public static string DefaultImage = "images/youtube.jpg";
        

        public static string BuildXmlName(string appName)
        {
            return appName + ".xml";
        }
        public static string BuildXmlName()
        {
            return DefaultAppName + ".xml";
        }


        public static string WriteLogFile(string basePath, string contents)
        {
            string time = ReplaceIllegalCharacters(DateTime.Now.ToString(), "_");
            string fileName = String.Format("UpdateLog_{0}.txt", time);
            string fullPath = Path.Combine(basePath, fileName);

            File.WriteAllText(fullPath, contents);
            return fullPath;    
        }

        public static string BuildFileName(string filePath, string fileName, FileType type)
        {
            switch (type)
            {
                case FileType.Image:
                    return Path.Combine(filePath, String.Format("{0}.jpg",fileName));
                case FileType.Flv:
                    return Path.Combine(filePath, String.Format("{0}.flv", FileHelper.ReplaceIllegalCharacters(fileName)));
                case FileType.Mp4:
                    return Path.Combine(filePath, String.Format("{0}.mp4", FileHelper.ReplaceIllegalCharacters(fileName)));
                case FileType.Wmv:
                    return Path.Combine(filePath, String.Format("{0}.wmv", FileHelper.ReplaceIllegalCharacters(fileName)));
                default:
                    return String.Empty;
                    
            }
        }

        public static double FileSize(string file)
        {
            FileInfo f = new FileInfo(file);
            double size = f.Length / 1024;
            return size;
        }

        public static string ReplaceIllegalCharacters(string source, string replacement)
        {

            if (!String.IsNullOrEmpty(replacement))
            {
                //make sure replacement doesn't have an illegal character
                replacement = ReplaceIllegalCharacters(replacement, String.Empty);
            }
            else
            {
                replacement = String.Empty;
            }

            //Using LINQ to replace characters for fun
            var query = from n in new[] { source }

                        //Replace illegal characters in DestinationFile name
                        // The following are illegal: \ / : * ? < > | " 
                        select n.Replace(@"\", replacement)
                                .Replace(@"/", replacement)
                                .Replace(":", replacement)
                                .Replace("*", replacement)
                                .Replace("?", replacement)
                                .Replace("<", replacement)
                                .Replace(">", replacement)
                                .Replace("|", replacement)
                                .Replace("\"", replacement) // quote char is escaped as \"
                                .Replace("-", replacement);

            return query.First();
        
        }

        public static string ReplaceIllegalCharacters(string source)
        {
            return ReplaceIllegalCharacters(source, String.Empty);
        }

        public static string BuildPath(ref string subDir)
        {
            //Environment.SpecialFolder doesn't have a folder for "Videos", 
            // but assume it does
            string newpath = Environment.GetFolderPath(Environment.SpecialFolder.MyMusic);
            OperatingSystem os = Environment.OSVersion;

            if (os.Version.Major >= 6) //Vista or higher
            {
                newpath = newpath.Replace("Music", "Videos");
            }
            else if (os.Version.Major == 5) //XP
            {
                newpath = newpath.Replace("My Music", "My Videos");
            }
            else
            {
                throw new NotSupportedException
                    (String.Format("The Operating System Version is not supported:{0}", Environment.OSVersion));                    
            }

            //Ensure the directory exists, if it doesn't, build it
            if (!Directory.Exists(newpath))
            {
                Directory.CreateDirectory(newpath);
            }

            if (String.IsNullOrEmpty(subDir))
            {
                subDir = Path.Combine(newpath, FileHelper.DefaultAppName);
            }
            else
            {
                subDir = Path.Combine(newpath, subDir);
            }           

            if (!Directory.Exists(subDir))
            {
                Directory.CreateDirectory(subDir);
            }               

            return newpath;                
        }

        public static bool ConvertedFilesExist(InnerTubeVideo video, Setting settings)
        {
            //iTunes needs WMV & MP4
            if (settings.iTunesInstalled)
            {
                if (File.Exists(video.DownloadedWmv) && File.Exists(video.DownloadedMp4))
                {
                    return true;
                }
                else
                {
                    return false; 
                }
            }
            else
            {
                if (File.Exists(video.DownloadedWmv))
                {
                    return true;
                }
                else
                {
                    return false; 
                }
            }
        }

    }
}

File: C04_InnerTube/InnerTubeVB/SharedUtilities/FileHelper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.IO

Namespace SharedUtilities
	Public Enum FileType
		Image
		Flv
		Mp4
		Wmv
	End Enum

	Public NotInheritable Class FileHelper
		Public Shared DefaultAppName As String = "InnerTube"
		Public Shared DefaultSettingsName As String = "settings.xml"
        Public Shared DefaultImage As String = "images/youtube.jpg"


		Private Sub New()
        End Sub

        Public Shared Function BuildXmlName(ByVal appName As String) As String
            Return appName & ".xml"
        End Function

        Public Shared Function BuildXmlName() As String
            Return DefaultAppName & ".xml"
        End Function

		Public Shared Function WriteLogFile(ByVal basePath As String, ByVal contents As String) As String
			Dim time As String = ReplaceIllegalCharacters(DateTime.Now.ToString(), "_")
			Dim fileName As String = String.Format("UpdateLog_{0}.txt", time)
			Dim fullPath As String = Path.Combine(basePath, fileName)

			File.WriteAllText(fullPath, contents)
			Return fullPath
		End Function

		Public Shared Function BuildFileName(ByVal filePath As String, ByVal fileName As String, ByVal type As FileType) As String
			Select Case type
				Case FileType.Image
					Return Path.Combine(filePath, String.Format("{0}.jpg",fileName))
				Case FileType.Flv
					Return Path.Combine(filePath, String.Format("{0}.flv", FileHelper.ReplaceIllegalCharacters(fileName)))
				Case FileType.Mp4
					Return Path.Combine(filePath, String.Format("{0}.mp4", FileHelper.ReplaceIllegalCharacters(fileName)))
				Case FileType.Wmv
					Return Path.Combine(filePath, String.Format("{0}.wmv", FileHelper.ReplaceIllegalCharacters(fileName)))
				Case Else
					Return String.Empty

			End Select
		End Function

		Public Shared Function FileSize(ByVal file As String) As Double
			Dim f As New FileInfo(file)
			Dim size As Double = f.Length / 1024
			Return size
		End Function

		Public Shared Function ReplaceIllegalCharacters(ByVal source As String, ByVal replacement As String) As String

			If (Not String.IsNullOrEmpty(replacement)) Then
				'make sure replacement doesn't have an illegal character
				replacement = ReplaceIllegalCharacters(replacement, String.Empty)
			Else
				replacement = String.Empty
			End If

			'Using LINQ to replace characters for fun
						'Replace illegal characters in DestinationFile name
						' The following are illegal: \ / : * ? < > | " 
            Dim query = From n In New String() {source} _
                        Select n.Replace("\", replacement) _
                        .Replace("/", replacement) _
                        .Replace(":", replacement) _
                        .Replace("*", replacement) _
                        .Replace("?", replacement) _
                        .Replace("<", replacement) _
                        .Replace(">", replacement) _
                        .Replace("|", replacement) _
                        .Replace("""", replacement) _
                        .Replace("-", replacement)
			Return query.First()

		End Function

		Public Shared Function ReplaceIllegalCharacters(ByVal source As String) As String
			Return ReplaceIllegalCharacters(source, String.Empty)
		End Function

		Public Shared Function BuildPath(ByRef subDir As String) As String
            'My.SpecialDirectories doesn't have a folder for "Videos", 
            ' but assume it exists

            Dim newPath As String = My.Computer.FileSystem.SpecialDirectories.MyMusic            
            Dim os As OperatingSystem = Environment.OSVersion

            If os.Version.Major >= 6 Then 'Vista or higher
                newPath = newPath.Replace("Music", "Videos")
            ElseIf os.Version.Major = 5 Then 'XP
                newPath = newPath.Replace("My Music", "My Videos")
            Else
                Throw New NotSupportedException(String.Format("The Operating System Version is not supported:{0}", Environment.OSVersion))
            End If

            'Ensure the directory exists, if it doesn't, build it
            If (Not Directory.Exists(newPath)) Then
                Directory.CreateDirectory(newPath)
            End If

            If String.IsNullOrEmpty(subDir) Then
                subDir = Path.Combine(newPath, FileHelper.DefaultAppName)
            Else
                subDir = Path.Combine(newPath, subDir)
            End If

            If (Not Directory.Exists(subDir)) Then
                Directory.CreateDirectory(subDir)
            End If

            Return newPath
		End Function

		Public Shared Function ConvertedFilesExist(ByVal video As InnerTubeVideo, ByVal settings As Setting) As Boolean
			'iTunes needs WMV & MP4
			If settings.iTunesInstalled Then
				If File.Exists(video.DownloadedWmv) AndAlso File.Exists(video.DownloadedMp4) Then
					Return True
				Else
					Return False
				End If
			Else
				If File.Exists(video.DownloadedWmv) Then
					Return True
				Else
					Return False
				End If
			End If
		End Function

	End Class
End Namespace
